addons/zylann.hterrain/native/src/quad_tree_lod.cpp

```diff
@@ -0,0 +1,248 @@
+}
+
+void QuadTreeLod::update(Vector3 view_pos) {
+	_update(ROOT, _max_depth, view_pos);
+
+	// This makes sure we keep seeing the lowest LOD,
+	// if the tree is cleared while we are far away
+	Quad *root = _get_root();
+	if (!root->has_children() && root->is_null())
+		root->set_data(_make_chunk(_max_depth, 0, 0));
+}
+
+void QuadTreeLod::debug_draw_tree(CanvasItem *ci) {
+	if (ci != nullptr)
+		_debug_draw_tree_recursive(ci, ROOT, _max_depth, 0);
+}
+
+// Intention is to only clear references to children
+void QuadTreeLod::_clear_children(unsigned int index) {
+	Quad *quad = _get_node(index);
+	if (quad->has_children()) {
+		_recycle_children(quad->first_child);
+		quad->first_child = NO_CHILDREN;
+	}
+}
+
+// Returns the index of the first_child. Allocates from _free_indices.
+unsigned int QuadTreeLod::_allocate_children() {
+	if (_free_indices.size() == 0) {
+		return NO_CHILDREN;
+	}
+
+	unsigned int i0 = _free_indices[_free_indices.size() - 1];
+	_free_indices.pop_back();
+	return i0;
+}
+
+// Pass the first_child index, not the parent index. Stores back in _free_indices.
+void QuadTreeLod::_recycle_children(unsigned int i0) {
+	// Debug check, there is no use case in recycling a node which is not a first child
+	CRASH_COND(i0 % 4 != 0);
+
+	for (int i = 0; i < 4; ++i) {
+		_node_pool[i0 + i].init();
+	}
+
+	_free_indices.push_back(i0);
+}
+
+Variant QuadTreeLod::_make_chunk(int lod, int origin_x, int origin_y) {
+	if (_make_func.is_valid()) {
+		return _make_func->call_func(origin_x, origin_y, lod);
+	} else {
+		return Variant();
+	}
+}
+
+void QuadTreeLod::_recycle_chunk(unsigned int quad_index, int lod) {
+	Quad *quad = _get_node(quad_index);
+	if (_recycle_func.is_valid()) {
+		_recycle_func->call_func(quad->get_data(), quad->origin_x, quad->origin_y, lod);
+	}
+}
+
+void QuadTreeLod::_join_all_recursively(unsigned int quad_index, int lod) {
+	Quad *quad = _get_node(quad_index);
+
+	if (quad->has_children()) {
+		for (int i = 0; i < 4; i++) {
+			_join_all_recursively(quad->first_child + i, lod - 1);
+		}
+		_clear_children(quad_index);
+
+	} else if (quad->is_valid()) {
+		_recycle_chunk(quad_index, lod);
+		quad->clear_data();
+	}
+}
+
+void QuadTreeLod::_update(unsigned int quad_index, int lod, Vector3 view_pos) {
+	// This function should be called regularly over frames.
+	Quad *quad = _get_node(quad_index);
+	int lod_factor = get_lod_factor(lod);
+	int chunk_size = _base_size * lod_factor;
+	Vector3 world_center = static_cast<real_t>(chunk_size) * (Vector3(static_cast<real_t>(quad->origin_x), 0.f, static_cast<real_t>(quad->origin_y)) + Vector3(0.5f, 0.f, 0.5f));
+
+	if (_vertical_bounds_func.is_valid()) {
+		Variant result = _vertical_bounds_func->call_func(quad->origin_x, quad->origin_y, lod);
+		Vector2 vbounds;
+		if (result.get_type() == Variant::VECTOR2)
+			vbounds = static_cast<Vector2>(result);
+		else
+			ERR_FAIL_COND("_vertical_bounds_func did not return a Vector2.");
```

Comment:
Invalid use of `ERR_FAIL_COND`: the way it works is to take the failing condition as argument.
So the code would have to be:
```cpp
Variant result = _vertical_bounds_func->call_func(quad->origin_x, quad->origin_y, lod);
ERR_FAIL_COND(result.get_type() != Variant::VECTOR2);
Vector2 vbounds = static_cast<Vector2>(result);
```
There is also a version taking an extra message if you wish:
```cpp
ERR_FAIL_COND_MSG(result.get_type() != Variant::VECTOR2, "Vertical bounds function must return a Vector2");
```
And when the passed condition is true, it internally does `return;`, aborting the function.
The compiler appears to have accepted it because technically a string is booleanizable to `true`, but that's a misuse of the macro nevertheless.

Reply:
Ah of course, my mistake. This has been fixed.